pkg/aggregator/aggregator.go

```diff
@@ -28,21 +27,42 @@ import (
 )
 
 // usedDefinitionForSpec returns a map with all used definitions in the provided spec as keys and true as values.
-func usedDefinitionForSpec(sp *spec.Swagger) map[string]bool {
+func usedDefinitionForSpec(root *spec.Swagger) map[string]bool {
 	usedDefinitions := map[string]bool{}
-	walkOnAllReferences(func(ref spec.Ref) spec.Ref {
-		if refStr := ref.String(); refStr != "" && strings.HasPrefix(refStr, definitionPrefix) {
-			usedDefinitions[refStr[len(definitionPrefix):]] = true
+	alreadyVisited := map[string]bool{}
+
+	walker := &readonlyReferenceWalker{
+		root: root,
+	}
+	walker.walkRefCallback = func(ref *spec.Ref) {
+		refStr := ref.String()
+		if refStr == "" || !strings.HasPrefix(refStr, definitionPrefix) {
+			return
 		}
-		return ref
-	}, sp)
+
+		defName := refStr[len(definitionPrefix):]
+		usedDefinitions[defName] = true
+
+		if _, found := root.Definitions[defName]; found && !alreadyVisited[refStr] {
+			alreadyVisited[refStr] = true
+			def := root.Definitions[defName]
+			walker.walkSchema(&def)
```

Comment:
I don't think the callback function should be responsible for this or even be able to call a part of walker.

Reply:
why not? I prefer to have this special logic cleanly separated. Note that we only do that in the readonly walker. The mutating walker does not need this and should not do this. Another sign, this is special behaviour.

Reply:
Why the mutating one do not need this?